monai/data/dataset.py

```diff
@@ -321,6 +323,238 @@ def __getitem__(self, index):
+        Update the cache items with new replacement for current epoch.
+
+        """
+        with self._update_lock:
+            if self._round_done:
+                remain_num: int = self.cache_num - self._replace_num
+                for i in range(remain_num):
+                    self._cache[i] = self._cache[i + self._replace_num]
+                for i in range(self._replace_num):
+                    self._cache[remain_num + i] = self._replacements[i]
+
+                self._start_pos += self._replace_num
+                if self._start_pos >= self._total_num:
+                    self._start_pos -= self._total_num
+
+                self._compute_data_idx()
+
+                # ready for next round
+                self._round += 1
+                self._round_done = False
+                return True
+            else:
+                return False
+
+    def update_cache(self):
+        """
+        Update cache items for current epoch, need to call this function before every epoch.
+        If the cache has been shutdown before, need to restart the `_replace_mgr` thread.
+
+        """
+        if not self._replace_mgr.isAlive():
+            self._restart()
+
+        # make sure update is done
+        while True:
+            if self._try_update_cache():
+                break
+            else:
+                time.sleep(0.01)
+
+    def _try_shutdown(self):
+        """
+        Wait for thread lock to shut down the background thread.
+
+        """
+        with self._update_lock:
+            if self._round_done:
+                self._round = 0
+                self._round_done = False
+                return True
+            else:
+                return False
+
+    def shutdown(self):
+        """
+        Shut down the background thread for replacement.
+
+        """
+        if not self._started:
+            return
+
+        # wait until replace mgr is done the current round
+        while True:
+            if self._try_shutdown():
+                self._replace_mgr.join()
+                return
+            else:
+                time.sleep(0.01)
+
+    def _replace_cache_thread(self, index: int):
+        """
+        Execute deterministic transforms on the new data for replacement.
+
+        """
+        pos: int = self._replace_data_idx[index]
+        self._replacements[index] = self._load_cache_item(self.data[pos], self.transform.transforms)  # type: ignore
+
+    def _compute_replacements(self):
+        """
+        Compute expected items for the replacement of next epoch, execute deterministic transforms.
+        It can support multi-threads to accelerate the computation progress.
+
+        """
+        if self.num_replace_workers > 0:
+            with ThreadPool(self.num_replace_workers) as p:
+                p.map(self._replace_cache_thread, list(range(self._replace_num)))
+        else:
+            for i in range(self._replace_num):
+                pos: int = self._replace_data_idx[i]
```

Comment:
This line and next could be replaced with `self._replace_cache_thread(i)`.

Reply:
yes, updated in the latest commit.
Thanks.